ruma-serde/src/canonical_json/value.rs

```diff
@@ -0,0 +1,248 @@
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!("a string").try_into().unwrap();
+    /// ```
+    String(String),
+
+    /// Represents a JSON array.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(["an", "array"]).try_into().unwrap();
+    /// ```
+    Array(Vec<CanonicalJsonValue>),
+
+    /// Represents a JSON object.
+    ///
+    /// The map is backed by a BTreeMap to guarantee the sorting of keys.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!({ "an": "object" }).try_into().unwrap();
+    /// ```
+    Object(BTreeMap<String, CanonicalJsonValue>),
+}
+
+impl CanonicalJsonValue {
+    pub fn to_canonical_string(&self) -> Result<String, serde_json::Error> {
+        Ok(serde_json::to_string(self).and_then(|s| {
+            if s.len() > 65_535 {
+                Err(serde_json::Error::custom("TOO LARGE"))
+            } else {
+                Ok(s)
+            }
+        })?)
+    }
+}
+impl fmt::Debug for CanonicalJsonValue {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            CanonicalJsonValue::Null => formatter.debug_tuple("Null").finish(),
+            CanonicalJsonValue::Bool(v) => formatter.debug_tuple("Bool").field(&v).finish(),
+            CanonicalJsonValue::Number(ref v) => fmt::Debug::fmt(v, formatter),
+            CanonicalJsonValue::String(ref v) => formatter.debug_tuple("String").field(v).finish(),
+            CanonicalJsonValue::Array(ref v) => {
+                formatter.write_str("Array(")?;
+                fmt::Debug::fmt(v, formatter)?;
+                formatter.write_str(")")
+            }
+            CanonicalJsonValue::Object(ref v) => {
+                formatter.write_str("Object(")?;
+                fmt::Debug::fmt(v, formatter)?;
+                formatter.write_str(")")
+            }
+        }
+    }
+}
+
+impl fmt::Display for CanonicalJsonValue {
+    /// Display a JSON value as a string.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// #
+    /// let json = json!({ "city": "London", "street": "10 Downing Street" });
+    ///
+    /// // Compact format:
+    /// //
+    /// // {"city":"London","street":"10 Downing Street"}
+    /// let compact = format!("{}", json);
+    /// assert_eq!(compact,
+    ///     "{\"city\":\"London\",\"street\":\"10 Downing Street\"}");
+    ///
+    /// // Pretty format:
+    /// //
+    /// // {
+    /// //   "city": "London",
+    /// //   "street": "10 Downing Street"
+    /// // }
+    /// let pretty = format!("{:#}", json);
+    /// assert_eq!(pretty,
+    ///     "{\n  \"city\": \"London\",\n  \"street\": \"10 Downing Street\"\n}");
+    /// ```
```

Comment:
Canonical JSON can't have a pretty format, otherwise it's not canonical JSON anymore.